File: articles/systems-biology-complexity-living-networks/c/systems_biology_kernel.c
/*
 * Compact systems biology kernel in C.
 */

#include <math.h>
#include <stdio.h>

void simulate_feedback(
    double x0,
    double y0,
    double production_x,
    double production_y,
    double degradation_x,
    double degradation_y,
    double hill_n,
    double dt,
    int steps,
    double *final_x,
    double *final_y
) {
    double x = x0;
    double y = y0;

    for (int step = 0; step < steps; step++) {
        double dx = production_x / (1.0 + pow(y, hill_n)) - degradation_x * x;
        double dy = production_y * x - degradation_y * y;

        x = x + dt * dx;
        y = y + dt * dy;

        if (x < 0.0) x = 0.0;
        if (y < 0.0) y = 0.0;
    }

    *final_x = x;
    *final_y = y;
}

int main(void) {
    double x;
    double y;

    simulate_feedback(0.20, 0.10, 1.20, 0.80, 0.40, 0.30, 2.0, 0.10, 80, &x, &y);

    printf("final_x=%.6f\n", x);
    printf("final_y=%.6f\n", y);

    return 0;
}

File: articles/systems-biology-complexity-living-networks/fortran/systems_biology_kernel.f90
! Compact systems biology feedback-dynamics kernel in Fortran.

program systems_biology_kernel
  implicit none

  integer, parameter :: steps = 80
  real :: x, y, dx, dy
  real :: production_x, production_y, degradation_x, degradation_y, hill_n, dt
  integer :: step

  x = 0.20
  y = 0.10
  production_x = 1.20
  production_y = 0.80
  degradation_x = 0.40
  degradation_y = 0.30
  hill_n = 2.0
  dt = 0.10

  do step = 1, steps
    dx = production_x / (1.0 + y**hill_n) - degradation_x * x
    dy = production_y * x - degradation_y * y

    x = x + dt * dx
    y = y + dt * dy

    if (x < 0.0) x = 0.0
    if (y < 0.0) y = 0.0
  end do

  print *, "final_x:", x
  print *, "final_y:", y
end program systems_biology_kernel
